clar2wasm/src/tools.rs

```diff
@@ -178,6 +181,108 @@ pub fn evaluate_at(
+            let mut processed_data = Vec::new();
+
+            for val in &list_data.data {
+                processed_data.push(unicode_to_byte_sequence(val.clone()));
+            }
+
+            Value::Sequence(SequenceData::List(ListData {
+                data: processed_data,
+                type_signature: list_data.type_signature.clone(),
+            }))
+        }
+        Value::Response(response_data) => Value::Response(ResponseData {
+            committed: response_data.committed,
+            data: Box::new(unicode_to_byte_sequence(*(response_data.data).clone())),
+        }),
+        Value::Optional(optional_data) => {
+            if let Some(data) = &optional_data.data {
+                Value::Optional(OptionalData {
+                    data: Some(Box::new(unicode_to_byte_sequence(*data.clone()))),
+                })
+            } else {
+                Value::Optional(optional_data.clone())
+            }
+        }
+        Value::Tuple(tuple_data) => {
+            let mut btree_map = BTreeMap::new();
+            for (name, value) in tuple_data.data_map.iter() {
+                btree_map.insert(name.clone(), unicode_to_byte_sequence(value.clone()));
+            }
+            Value::Tuple(TupleData {
+                data_map: btree_map,
+                type_signature: tuple_data.type_signature.clone(),
+            })
+        }
+        Value::Sequence(SequenceData::String(CharType::UTF8(u))) => {
+            println!("Input Data :{:?}", u.data);
+
+            let scalar_bytes = u.data.clone();
+
+            let mut utf8_bytes: Vec<Vec<u8>> = Vec::new();
+
+            for bytes in scalar_bytes {
+                // Reconstruct the scalar value assuming big-endian byte order
+                let mut scalar_value: u32 = 0;
+                let num_bytes = bytes.len();
+                for (i, &byte) in bytes.iter().enumerate() {
+                    scalar_value |= (byte as u32) << (8 * (num_bytes - 1 - i));
+                }
+                println!("Scalar value: {scalar_value}");
+                // Convert the scalar value to a char if it's a valid Unicode scalar value
+                if let Some(character) = std::char::from_u32(scalar_value) {
+                    if character.is_ascii() {
+                        utf8_bytes.push(vec![character as u8]);
+                    } else {
+                        let mut buf = [0; 4];
+                        let encoded_str = character.encode_utf8(&mut buf);
+                        utf8_bytes.push(encoded_str.as_bytes().to_vec());
+                    }
+                }
+            }
+
+            println!("Output UTF-8 Data: {:?}", utf8_bytes);
+
+            Value::Sequence(SequenceData::String(CharType::UTF8(UTF8Data {
+                data: utf8_bytes,
+            })))
+        }
+        _ => input,
+    }
+}
+
+pub fn unicode_scalars_from_string(str: String) -> Value {
+    let validated_utf8_str = std::str::from_utf8(str.as_bytes()).unwrap();
```

Comment:
What's the purpose of this line? Aren't rust strings guaranteed to be valid utf8?

Reply:
Yep. That's kind of redundant. I was just testing it